### PR TITLE
Don't create an image layer for dropped relations.

### PR DESCRIPTION
**Pull request by hlinnaka**
I noticed that the timeline directory contained files like this:

    pg_xact_0000_0_000000000169C3C2_00000000016BB399
    pg_xact_0000_0_00000000016BB399
    pg_xact_0000_0_00000000016BB399_00000000016BDD06
    pg_xact_0000_0_00000000016BDD06
    pg_xact_0000_0_00000000016BDD06_00000000016C63AA
    pg_xact_0000_0_00000000016C63AA
    pg_xact_0000_0_00000000016C63AA_0000000001765226_DROPPED
    pg_xact_0000_0_0000000001765226
    pg_xact_0001_0_00000000016BB77E_00000000016BDD06
    pg_xact_0001_0_00000000016BDD06
    pg_xact_0001_0_00000000016BDD06_0000000001765226_DROPPED
    pg_xact_0001_0_0000000001765226

Note how there is an image file after each DROPPED file. It's a waste of
time and space to materialize an image of the file at the point where it's
dropped, no one is going to request pages on a dropped relation. And it's
a correctness issue too: list_rels() and list_nonrels() will not consider
the relation as unlinked, unless the latest layer indicates so, and there
is no concept of a dropped image layer. That was causing test_clog_truncate
test to fail, when I adjusted the checkpointer to force a checkpoint more
aggressively.

There are a bunch more issues related to dropped rels and branching,
see https://github.com/zenithdb/zenith/issues/502. Hence this doesn't
completely fix the issue I saw with test_clog_truncate either. But it's
a start.

### CODE REVIEW

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 1]
```diff
@@ -522,29 +516,36 @@ impl InMemoryLayer {
...
+        if !dropped {
+            // Write a new base image layer at the cutoff point
+            let imgfile = ImageLayer::create_from_src(self.conf, timeline, self, end_lsn)?;
+            let imgfile_rc: Arc<dyn Layer> = Arc::new(imgfile);
+            frozen_layers.push(Arc::clone(&imgfile_rc));
+            trace!("freeze: created image layer {} at {}", self.seg, end_lsn);
+
+            // If there were any page versions newer than the cutoff, initialize a new in-memory
+            // layer to hold them
+            if !dropped && (!after_segsizes.is_empty() || !after_page_versions.is_empty()) {
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:529]: `!droppped` is redundant

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 2]
```diff
@@ -522,29 +516,36 @@ impl InMemoryLayer {
...
-            Some(Arc::new(new_open))
-        } else {
-            None
-        };
+        let mut new_open_rc = None;
+        if !dropped {
+            // Write a new base image layer at the cutoff point
+            let imgfile = ImageLayer::create_from_src(self.conf, timeline, self, end_lsn)?;
+            let imgfile_rc: Arc<dyn Layer> = Arc::new(imgfile);
+            frozen_layers.push(Arc::clone(&imgfile_rc));
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:524]: Does the order that layers are pushed to `frozen_layers` matter? Before, the image layer was pushed earlier than the delta, now it's reversed.

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 3]
```diff
@@ -522,29 +516,36 @@ impl InMemoryLayer {
...
-            new_inner.segsizes.append(&mut after_segsizes);
-            drop(new_inner);
-            trace!("freeze: created new in-mem layer {} {}-", self.seg, end_lsn);
-
-            Some(Arc::new(new_open))
-        } else {
-            None
-        };
+        let mut new_open_rc = None;
+        if !dropped {
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:520]: Optional, but alternative style: `let new_open_rc = if !dropped { .. } else { None };`

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 2 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/inmemory_layer.rs:524]: True, but doesn't matter.

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 1 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/inmemory_layer.rs:529]: Fixed, thanks

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 3 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/inmemory_layer.rs:520]: It's a bit awkward when there are multiple branches that lead to None. It looked like this:
```
let new_open_rc = if !dropped {
    ...
    if !after_segsizes.is_empty() || !after_page_versions.is_empty() {
        ...
        Some(Arc::new(new_open))
    } else {
        None
    }
} else {
    None
}
```

Comment (patins) [reply pageserver/src/layered_repository/inmemory_layer.rs:520]: Ahh I see

**Review by patins — APPROVED** [review]

**Pull request merged**